media-proxy/src/session-mtl-st2110-20.cc

```diff
@@ -0,0 +1,416 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "session-mtl.h"
+
+#if defined(MTL_ZERO_COPY)
+static int query_ext_frame_callback_wrapper(void *priv, struct st_ext_frame *ext_frame,
+                                            struct st20_rx_frame_meta *meta)
+{
+    rx_st20_mtl_session *s = (rx_st20_mtl_session *)priv;
+    return s->query_ext_frame_cb(ext_frame, meta);
+}
+
+static int tx_frame_done_callback_wrapper(void *priv, struct st_frame *frame)
+{
+    tx_st20_mtl_session *s = (tx_st20_mtl_session *)priv;
+    return s->frame_done_cb(frame);
+}
+#endif
+
+void rx_st20_mtl_session::frame_thread()
+{
+    INFO("%s, start\n", __func__);
+    while (!stop) {
+        struct st_frame *frame = st20p_rx_get_frame(handle);
+        if (!frame) { /* no frame */
+            pthread_mutex_lock(&wake_mutex);
+            if (!stop)
+                pthread_cond_wait(&wake_cond, &wake_mutex);
+            pthread_mutex_unlock(&wake_mutex);
+            continue;
+        }
+        /* Verify frame data status. */
+        if (frame->status == ST_FRAME_STATUS_CORRUPTED) {
+            ERROR("[DBG] Received corrupted frame.\n");
+        } else {
+            consume_frame(frame);
+        }
+        st20p_rx_put_frame(handle, frame);
+    }
+}
+
+/* Create new RX session */
+rx_st20_mtl_session::rx_st20_mtl_session(mtl_handle dev_handle, struct st20p_rx_ops *opts,
+                                         memif_ops_t *memif_ops)
+    : handle(0), frame_thread_handle(0), shm_bufs(0), fb_recv(0)
+{
+    if (!dev_handle || !opts || !memif_ops) {
+        cleanup();
+        throw invalid_argument("Input parameters cannot be null");
```

Comment:
I have a strong suggestion not to throw exceptions but return errors on all layers of the calling stack and process return values explicitly.

Reply:
Constructor cannot return error. The only way to handle error path nicely is to use throw

Reply:
You can extract the functionality that can fail in an `init()` method which returns an error.
The code could look as below
```c
A *a = new A;
int err = a->init();
if (err) {
    delete a;
    return err;
}
```
Simple. Easy to read.

Reply:
I totally agree that we have to handle errors in places where they occur, but I'm not that sure that doing that by return values is significantly better then handling exceptions.

If we chose to not throw in constructors, i would do something like this, and make the constructor private:
```c++
class A{
public:
    static std::optional<A> create(size_t size) {
        A obj;
        if (obj.initialize(size)) {
            return obj;  // Return constructed object
        }
        return std::nullopt;  // Indicate failure to construct
    }


private:
    A() = default;

    bool initialize(size_t size) {
        try {
            data = new int[size];  // May throw std::bad_alloc
            return true;
        } catch (const std::bad_alloc&) {
            std::cerr << "Memory allocation failed in initialize." << std::endl;
            return false;
        }
    }

    ~A() {
        delete[] data;
    }

    int* data = nullptr;
};

int main() {
    size_t size = 1000000;  // Large size for demonstration

    // Try to create the A using the factory method
    auto objOpt = A::create(size);

    if (objOpt) {
        // Object was successfully created, use it
    } else {
        // Handle the failure to create the object
        std::cerr << "Failed to create A due to memory allocation issues." << std::endl;
    }
```

Reply:
Exactly. In this example you don't throw your own exceptions. However, you catch exceptions that can occur in a 3rd party component. And then you return an error. That's what I want to achieve.

I suggest we establish such a strategy of three statements:
* Don't throw own exceptions.
* Always catch 3rd party component exceptions.
* Always handle and return errors.

Reply:
Another **huge** drawback of implementing some functionality in a constructor is that you can't call overridden virtual methods in the constructor. It means the base class can't call a virtual method implemented in the derived class. The only option is to only initialize the class instance in the constructor, and when the instance is created, call a virtual method to do what your feature should do.

Reply:
1. Throw removed, 
2. mtl/rdma/memif initialization moved to init() method